### ISSUE TITLE
Rule proposal: `no-unnecessary-type-casting`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#618]
# FAIL

```js
if (Boolean(foo)) {}
```

```js
if (!!foo) {}
```

```js
foo(Boolean(bar) ? 1 : 2)
```

```js
const foo = Boolean(bar);
if (foo) {}
```

Not sure about the last one, the example seems need fail, but [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/2e273e9830bfda68a7327fa614fee205c90fd9a1/rules/expiring-todo-comments.js#L21) and [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/cea134612a7be8de1ce91082caba5ce4d73166b2/rules/number-literal-case.js#L17) seems more readable with type-casting.

# PASS

```js
if (foo) {}
```

```js
foo(bar ? 1 : 2)
```

---

Got the idea from #616 [L167](https://github.com/sindresorhus/eslint-plugin-unicorn/commit/2146e5a2fd18138921b29e7646e5d88992ac965c#diff-dd6f3ff5cba5d13fe6025d9423e81447L167) [L195](https://github.com/sindresorhus/eslint-plugin-unicorn/commit/2146e5a2fd18138921b29e7646e5d88992ac965c#diff-dd6f3ff5cba5d13fe6025d9423e81447L195), examples only shows `Boolean` type-casting, haven't thought about other types.

**Comment by himynameisdave** [comment]
Is this similar to [this rule in `@typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unnecessary-boolean-literal-compare.md)?

Do we need type information to make this rule work, or is it more just ensuring that we're not doing an extra boolean cast in conditional statements?

**Comment by papb** [comment]
> const foo = Boolean(bar);
> if (foo) {}

I think this has to be a pass... If it fails, what about the following examples?

```js
// Example 1: this must pass
function bar(baz) {
  const foo = Boolean(baz);
  if (foo) {
    console.log('Truthy input!');
  }
  return foo;
}

// Example 2: this must pass
function bar(baz) {
  const foo = Boolean(baz);
  baz = whatever();
  if (foo) {
    console.log('Truthy input!');
  }
}
```

**Comment by fisker** [comment]
@papb  Sorry, I missed details, we should check `foo` is only in `if` or `ternary`.

So

```js
function bar(baz) {
  const foo = Boolean(baz);
  if (foo) {}
  return foo;  // any use of foo will pass
}
```

**Comment by papb** [comment]
@fisker Yeah, that solves my example 1, but what about example 2?

**Comment by fisker** [comment]
@papb Right, I didn't thought about that before, but  we can just remove `Boolean`, right ?

**Comment by papb** [comment]
Maybe... But I think it's getting too complicated to ensure the rule does not claim an error where it's not;

Also, back to the other point, as you said:

> [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/2e273e9830bfda68a7327fa614fee205c90fd9a1/rules/expiring-todo-comments.js#L21) and [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/cea134612a7be8de1ce91082caba5ce4d73166b2/rules/number-literal-case.js#L17) seems more readable with type-casting.

And I fully agree. So let's stick with operations directly performed inside an if condition or ternary operator...

**Comment by sindresorhus** [comment]
Accepted.

**Comment by fisker** [comment]
`no-unnecessary-type-casting` or just `no-unnecessary-boolean-type-casting`?

Not really have any useful cases for other type in my mind.


```js
for (const foo of [...iterable]) {} // Saw people use this.
const foo = '' + `${bar}`
const foo = String(`${bar}`)
```

They both not seen too much

**Comment by sindresorhus** [comment]
`no-unnecessary-boolean-coercion` ?

**Comment by papb** [comment]
`no-unnecessary-boolean-casting` ?

**Comment by papb** [comment]
@fisker Is `for (const foo of [...iterable])` entirely equivalent to `for (const foo of iterable)` though? What if the loop alters the iterable in some way? (I never alter iterables inside the loop itself so I don't know how it would behave)

**Comment by fisker** [comment]
@papb I was trying to find some other case not about `boolean`, point is some people thought `for..of` only work on array.

Do you have any unnecessary type casting case in your mind?

**Comment by papb** [comment]
@fisker None... If I think of any I'll post here.

**Comment by fisker** [comment]
I was going to implement this rule. But found this rule just before I get started.

[The the built-in ESLint `no-extra-boolean-cast` rule](https://eslint.org/docs/rules/no-extra-boolean-cast).

Seems it's doing the job I propose.
